go-controller/hybrid-overlay/pkg/controller/node_linux.go

```diff
@@ -53,26 +54,62 @@ func podToCookie(pod *kapi.Pod) string {
 	return nameToCookie(pod.Namespace + "_" + pod.Name)
 }
 
-func (n *NodeController) addOrUpdatePod(pod *kapi.Pod) error {
+func (n *NodeController) waitForNamespace(name string, wf *factory.WatchFactory) (*kapi.Namespace, error) {
+	var namespaceBackoff = wait.Backoff{Duration: 1 * time.Second, Steps: 7, Factor: 1.5, Jitter: 0.1}
+	var namespace *kapi.Namespace
+	if err := wait.ExponentialBackoff(namespaceBackoff, func() (bool, error) {
+		var err error
+		namespace, err = wf.GetNamespace(name)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				// Namespace not found; don't bother waiting longer
+				return false, err
+			}
+			klog.Warningf("error getting namespace: %v", err)
+			return false, nil
+		}
+		return true, nil
+	}); err != nil {
+		return nil, fmt.Errorf("failed to get namespace object: %v", err)
+	}
+	return namespace, nil
+}
+
+func (n *NodeController) addOrUpdatePod(pod *kapi.Pod, wf *factory.WatchFactory) error {
 	podIPs, podMAC, err := getPodDetails(pod, n.nodeName)
 	if err != nil {
 		klog.V(5).Infof("cleaning up hybrid overlay pod %s/%s because %v", pod.Namespace, pod.Name, err)
 		return n.deletePod(pod)
 	}
 
+	namespace, err := n.waitForNamespace(pod.Namespace, wf)
+	if err != nil {
+		return err
+	}
+	namespaceExternalGW := namespace.GetAnnotations()[hotypes.HybridOverlayExternalGw]
 	cookie := podToCookie(pod)
 	for _, podIP := range podIPs {
 		_, _, err = util.RunOVSOfctl("add-flow", extBridgeName,
 			fmt.Sprintf("table=10,cookie=0x%s,priority=100,ip,nw_dst=%s,actions=set_field:%s->eth_src,set_field:%s->eth_dst,output:ext", cookie, podIP.IP, n.drMAC, podMAC))
 		if err != nil {
 			return fmt.Errorf("failed to add flows for pod %s/%s: %v", pod.Namespace, pod.Name, err)
 		}
+		if namespaceExternalGW != "" {
+			_, _, err = util.RunOVSOfctl("add-flow", extBridgeName,
+				fmt.Sprintf("table=0, cookie=0x%s, priority=100, ip, nw_src=%s, actions=load:%d->NXM_NX_TUN_ID[0..31],set_field:%s->tun_dst,output:%s", cookie, podIP.IP, hotypes.HybridOverlayVNI, namespaceExternalGW, extVXLANName))
```

Comment:
Let's discuss this flow tmrw. I think we may need to make it more specific on the match criteria.

Reply:
@rcarrillocruz you should add detection for ip type here for the pod , and match on ipv6

Reply:
i thought the pods at this point would be ipv4, as the 6to4 was performed by the external gw.
@dcbw can you pls comment?

Reply:
Hybrid overlay doesn't support v6 yet, we need to make that happen. Can address in a follow-up I think.